examples/tox21/gcn_sparse_pattern_example.py

```diff
@@ -0,0 +1,193 @@
+    # training params
+    lr = args.lr
+    num_epochs = args.epochs
+    batch_size = args.batch_size
+    task = args.task
+    early_stop_patience = args.early_stop
+
+    # setup model
+    def forward(node_feats: jnp.ndarray, adj: jnp.ndarray, graph_idx: jnp.ndarray,
+                is_training: bool) -> jnp.ndarray:
+        """Forward application of the GCN."""
+        model = GCNPredicator(in_feats=in_feats, hidden_feats=hidden_feats, activation=activation,
+                              batch_norm=batch_norm, dropout=dropout, pooling_method=pooling_method,
+                              predicator_hidden_feats=predicator_hidden_feats,
+                              predicator_dropout=predicator_dropout, n_out=n_out)
+        preds = model(node_feats, adj, graph_idx, is_training)
+        logits = clipped_sigmoid(preds)
+        return logits
+
+    model = hk.transform_with_state(forward)
+    optimizer = optix.adam(learning_rate=lr)
+
+    # define training loss
+    def train_loss(params: hk.Params, state: State, batch: Batch,
+                   graph_idx: jnp.ndarray) -> Tuple[jnp.ndarray, State]:
+        """Compute the loss."""
+        inputs, targets = batch
+        logits, new_state = model.apply(params, state, next(rng_seq), *inputs, graph_idx, True)
+        loss = binary_cross_entropy(logits, targets)
+        return loss, new_state
+
+    # define training update
+    @partial(jax.jit, static_argnums=(4,))
+    def update(params: hk.Params, state: State, opt_state: OptState,
+               batch: Batch, graph_idx: jnp.ndarray) -> Tuple[hk.Params, State, OptState]:
+        """Update the params."""
+        (_, new_state), grads = jax.value_and_grad(train_loss, has_aux=True)(params, state, batch, graph_idx)
+        updates, new_opt_state = optimizer.update(grads, opt_state)
+        new_params = optix.apply_updates(params, updates)
+        return new_params, new_state, new_opt_state
+
+    # define evaluate metrics
+    @partial(jax.jit, static_argnums=(3,))
+    def evaluate(params: hk.Params, state: State, batch: Batch,
+                 graph_idx: jnp.ndarray) -> jnp.ndarray:
+        """Compute evaluate metrics."""
+        inputs, targets = batch
+        logits, _ = model.apply(params, state, next(rng_seq), *inputs, graph_idx, False)
+        loss = binary_cross_entropy(logits, targets)
+        return logits, loss, targets
+
+    print("Starting training...")
+    task_index = tox21_tasks.index(task)
+    early_stop = EarlyStopping(patience=early_stop_patience)
+    init_batch, init_graph_idx = collate_fn(
+        next(train_dataset.iterbatches(batch_size=batch_size)), task_index
+    )
+    params, state = model.init(next(rng_seq), *init_batch[0], init_graph_idx, True)
+    opt_state = optimizer.init(params)
+    for epoch in range(num_epochs):
+        # train
+        start_time = time.time()
+        # FIXME : This for loop should be rewrited by lax.scan or lax.fori_loop
```

Comment:
Please confirm the details from this issue : https://github.com/google/jax/issues/2242
SparseCGN uses `jax.ops.index_add`, but a large Python "for" loop leads to a serious performance issue when using `jax.ops.index_add` (training time of this example is 24 times than the example of PadGCN)

Currently, I try to fix this problem.

Reply:
Good find. Maxes sense!